readthedocs/analytics/utils.py

```diff
@@ -0,0 +1,107 @@
+"""Utilities related to analytics"""
+
+from __future__ import absolute_import, unicode_literals
+import hashlib
+import logging
+
+from django.conf import settings
+from django.utils.encoding import force_text, force_bytes
+from django.utils.crypto import get_random_string
+import requests
+from user_agents import parse
+
+try:
+    # Python 3.3+ only
+    import ipaddress
+except ImportError:
+    from .vendor import ipaddress
+
+log = logging.getLogger(__name__)   # noqa
+
+# Used to anonymize an IP by zero-ing out the last 2 bytes
+MASK = int('0xFFFFFFFFFFFFFFFFFFFFFFFFFFFF0000', 16)
+
+
+def get_client_ip(request):
+    """Gets the real IP based on a request object"""
+    ip_address = request.META.get('REMOTE_ADDR')
+
+    # Get the original IP address (eg. "X-Forwarded-For: client, proxy1, proxy2")
+    x_forwarded_for = request.META.get('HTTP_X_FORWARDED_FOR', '').split(',')[0]
+    if x_forwarded_for:
+        ip_address = x_forwarded_for
+
+    return ip_address
+
+
+def anonymize_ip_address(ip_address):
+    """Anonymizes an IP address by zeroing the last 2 bytes"""
+    try:
+        ip_obj = ipaddress.ip_address(force_text(ip_address))
+    except ValueError:
+        return None
+
+    anonymized_ip = ipaddress.ip_address(int(ip_obj) & MASK)
+    return anonymized_ip.compressed
+
+
+def anonymize_user_agent(user_agent):
+    """Anonymizes rare user agents"""
+    # If the browser family is not recognized, this is a rare user agent
+    parsed_ua = parse(user_agent)
+    if parsed_ua.browser.family == 'Other' or parsed_ua.os.family == 'Other':
+        return 'Rare user agent'
+
+    return user_agent
+
+
+def send_to_analytics(data):
+    """Sends data to Google Analytics"""
+    if data.get('uip') and data.get('ua'):
+        data['uid'] = generate_client_id(data['uip'], data['ua'])
+
+    if 'uip' in data:
+        # Anonymize IP address if applicable
+        data['uip'] = anonymize_ip_address(data['uip'])
+
+    if 'ua' in data:
+        # Anonymize user agent if it is rare
+        data['ua'] = anonymize_user_agent(data['ua'])
+
+    resp = None
+    try:
+        resp = requests.post(
+            'https://www.google-analytics.com/collect',
+            data=data,
+        )
+    except requests.Timeout:
```

Comment:
What is the default timeout here? We should probably set it to something really low.